### ISSUE TITLE
Fonts with variants are rendered wrong

### ISSUE THREAD

**Issue by arudp** [microsoft/vscode#105672]
- VSCode Version: 1.48.2
- OS Version: macOS Catalina 10.15.6 19G73 x86_64

Steps to Reproduce:
1. Use IBM Plex Mono or a custom built Iosevka. Probably happens with other fonts.

Does this issue occur when all extensions are disabled?: Yes

---

As you can see in the following screenshots, at least these fonts in particular are rendered wrong in the VSCode editor. Strangely enough, the terminal embedded in VSCode renders them just fine. I'm wondering if this might be a VSCode issue or if I could be using them wrong, though I haven't found this to be an issue in any other editor or in any browser, including Chrome. 

<img width="1528" alt="Screenshot 2020-08-30 at 11 12 01" src="https://user-images.githubusercontent.com/44339057/91655547-b0877600-eab1-11ea-89ac-f2033a45f7be.png">

<img width="1528" alt="Screenshot 2020-08-30 at 11 09 35" src="https://user-images.githubusercontent.com/44339057/91655507-55558380-eab1-11ea-8181-7f54b4c25d89.png">

I found this issue in Linux, MacOS and Windows. I guess the issue with Iosevka is somehow connected with how customizable the font is and the variants you can choose from, and maybe IBM Plex Mono too? The thing is, this doesn't seem to happen with Consolas, JetBrains Mono, Source Code Pro or any other fonts that I've tried that have different rendering for italics.

**Comment by alexdima** [comment]
@arudp I'm not familiar with IBM Plex Mono or Iosevka. Could you please show the minimal relevant settings, and how you expect the font to render vs what it renders? I can then try to reproduce on my machine, but as it is right now, I don't quite understand what the screenshots are supposed to show.

**Comment by arudp** [comment]
@alexdima Sure, my bad for not explaining properly. Actually the screenshots have both those things. As you can see, in one screenshot the font settings are set to be IBM Plex Mono and the other ones to Iosevka

```json
"editor.fontFamily": "Iosevka-Code-Extended,CascadiaCode ... more fonts",
"debug.console.fontFamily": "Iosevka-Code-Extended, ... ",
```

The part to see the difference is a little trickier, so I should've explained myself better, sorry. In both screenshots VSCode is split with editor (Upper half) and terminal (Lower half). The editor seems to have trouble rendering the fonts, while the terminal renders them just perfectly. 

In the IBM Plex Mono screenshot (first one) you can see that in the editor the letter "a" is show different when in italics, right? But it's showing them the other way around, the "a" it shows in regular style should be the one shown when the text is italics and viceversa.

In Iosevka though there are a lot more changes. [Iosevka](https://github.com/be5invis/Iosevka/) is a really cool font that lets you decide the variants for a lot of characters. 
In the terminal you can see the font as I built it with the styles and script they provide, but as you can see in the editor, they are quite different (for example, "i", "r", "g", "k" characters).

**Comment by alexdima** [comment]
@arudp Thanks for the prompt reply. It is possible that the terminal renders things differently than the editor. xterm.js uses custom rendering, where they sample glyphs and create GPU textures / atlases and then render the terminal. This mostly works because of some monospace assumptions. The editor on the other hand delegates everything to Chromium.

But in any case, I've install iosevka on my machine and have configured the following relevant settings:
```
	"editor.fontFamily": "iosevka",
	"editor.tokenColorCustomizations": {
		"textMateRules": [
			{ "scope": "comment", "settings": { "fontStyle": "italic" } }
		]
	},
	"workbench.colorTheme": "Default Light+",
```

I cannot notice any difference in rendering between the editor and the terminal:
![image](https://user-images.githubusercontent.com/5047891/92042144-d8a8fa80-ed79-11ea-92cf-66f5760221a6.png)

**Comment by arudp** [comment]
@alexdima Right, but that's because you installed the default variant of Iosevka. If you look closely you'll notice that my editor Iosevka looks very much like the Iosevka you just installed, but even when the terminal uses the same setting to choose the font, you can see that the terminal font looks differently.

I guessed that at least one of those were leaving things to Chromium, but I don't see this issue in Chrome, might it be that VSCode uses an older version of Chrome that the one that's currently out?

I'm attaching my version of Iosevka just in case you want to try, the setting is, if you're in Mac, `Iosevka-Code-Extended`. If in Linux or Windows I think instead of the dashes you can just use whitespace if I'm not wrong, though I'm right now on my Mac and cannot confirm the other OSs, sorry.

[iosevkaCodeExtended.zip](https://github.com/microsoft/vscode/files/5165114/iosevkaCodeExtended.zip)

**Comment by alexdima** [comment]
@arudp I see you have configured `"editor.fontLigatures": "'ss01, 'ss05', 'dlig'"`. Have you tried enabling ligatures via `"editor.fontLigatures": "'ss01, 'ss05', 'dlig', 'liga' on, 'calt' on"` or commenting out that setting? (that setting influences font rendering in the editor, but not in the integrated terminal).

**Comment by arudp** [comment]
@alexdima That was exactly it! Damn, I completely forgot about that setting, so silly!! Thank you very much!

**Comment by spinicist** [comment]
@alexdima Thanks so much for this solution. Could you explain the `fontLigatures` setting, or where to get more information on it please?

I use Source Code Pro on a Mac, and today hit the problem that `l l` (note the space) was being rendered like this:
![Screenshot 2020-09-14 at 15 36 48](https://user-images.githubusercontent.com/11050744/93100630-4b32a800-f6a1-11ea-8197-2e927f294df8.png)

Adding the `fontLigatures` setting has solved this, but I don't understand why.

**Comment by alexdima** [comment]
Under the hood, we transform the setting `editor.fontLigatures` to a CSS property named `font-feature-settings`. If `editor.fontLigatures` is:
* `false` => we set `style.fontFeatureSettings = '"liga" off, "calt" off'`.
* `true` => we set `style.fontFeatureSettings = '"liga" on, "calt" on'`.
* a string => we set `style.fontFeatureSettings = '${user_provided_string}'`.

It looks like with our latest stable release, which upgrades our Electron version from Electron 7 to Electron 9, we also shipped a new layouting subsystem from Chromium called `layoutNG`. It looks like this new sybsystem has trouble understanding  `"liga" off, "calt" off`, but will understand just fine `'liga' off, 'calt' off`. I am investigating this at https://github.com/microsoft/vscode/issues/106583
